pdq/cpp/bin/benchmark-query.cpp

```diff
@@ -0,0 +1,323 @@
+          usage(argv0, 1);
+          return;
+        }
+      }
+    } else if (arg == "-v") {
+      verbose = true;
+    } else if (arg == "-h" || arg == "--help") {
+      usage(argv0, 0);
+      return;
+    } else if (arg.length() > 0) {
+      fprintf(stderr, "Unknown argument: %s\n", arg.c_str());
+      usage(argv0, 1);
+      return;
+    }
+  }
+
+  // Initialize random number generator
+  std::mt19937 gen(seed);
+
+  // Generate random hashes for queries
+  std::vector<std::pair<facebook::pdq::hashing::Hash256, std::string>> queries;
+  for (size_t i = 0; i < querySize; i++) {
+    auto hash = generateRandomHash(gen);
+    queries.push_back({hash, "query_" + std::to_string(i)});
+  }
+
+  // Generate random hashes for index
+  std::vector<std::pair<facebook::pdq::hashing::Hash256, std::string>> index;
+  for (size_t i = 0; i < indexSize - querySize; i++) {
+    auto hash = generateRandomHash(gen);
+    index.push_back({hash, "index_" + std::to_string(i)});
+  }
+
+  // Add noise to queries then insert into index
+  std::uniform_int_distribution<int> noiseDist(1, maxDistance);
+  for (const auto& query : queries) {
+    int bitsToFlip = noiseDist(gen);
+    auto noisyHash = addNoise(query.first, bitsToFlip, gen);
+    index.push_back({noisyHash, "index_noisy_" + query.second});
+  }
+  std::shuffle(index.begin(), index.end(), gen);
+
+  if (verbose) {
+    printf("GENERATED QUERIES:\n");
+    for (const auto& it : queries) {
+      printf("%s,%s\n", it.first.format().c_str(), it.second.c_str());
+    }
+    printf("\n");
+
+    printf("GENERATED INDEX:\n");
+    for (const auto& it : index) {
+      printf("%s,%s\n", it.first.format().c_str(), it.second.c_str());
+    }
+    printf("\n");
+  }
+
+  if (method == "linear") {
+    queryLinear(
+        maxDistance, verbose, seed, indexSize, querySize, queries, index);
+  } else if (method == "mih") {
+    queryMIH(maxDistance, verbose, seed, indexSize, querySize, queries, index);
+  }
+}
+
+///////////////////////
+//// Query methods ////
+///////////////////////
+
+static void queryLinear(
+    const int maxDistance,
+    const bool verbose,
+    const unsigned int seed,
+    const size_t indexSize,
+    const size_t querySize,
+    const std::vector<std::pair<facebook::pdq::hashing::Hash256, std::string>>&
+        queries,
+    const std::vector<std::pair<facebook::pdq::hashing::Hash256, std::string>>&
+        index) {
+  // Do linear searches
+  std::vector<std::pair<facebook::pdq::hashing::Hash256, std::string>> matches;
+
+  std::chrono::time_point<std::chrono::steady_clock> t1, t2;
+  std::chrono::duration<double> elapsedSeconds;
+
+  t1 = std::chrono::steady_clock::now();
+  for (const auto& it : queries) {
+    for (const auto& it2 : index) {
+      if (it.first.hammingDistance(it2.first) <= maxDistance) {
+        matches.push_back(it2);
+      }
+    }
+  }
+  t2 = std::chrono::steady_clock::now();
+  elapsedSeconds = t2 - t1;
+  double seconds = elapsedSeconds.count();
```

Comment:
the mih has an indexing step which takes some time. i'm not sure how to refactor it cleanly so that the index doesn't get included in the query timing. how would you suggest i do this?
```cpp
  // Build the MIH
  std::chrono::time_point<std::chrono::steady_clock> t1, t2;
  std::chrono::duration<double> elapsedSeconds;
  double indexTimeSeconds;

  facebook::pdq::index::MIH256<std::string> mih;

  t1 = std::chrono::steady_clock::now();
  for (const auto& it : index) {
    mih.insert(it.first, it.second);
  }
  t2 = std::chrono::steady_clock::now();
  elapsedSeconds = t2 - t1;
  indexTimeSeconds = elapsedSeconds.count();
  printf("\n");
  if (verbose) {
    printf("\n");
    mih.dump();
    printf("\n");
  }

  // Do indexed searches
  std::vector<std::pair<facebook::pdq::hashing::Hash256, std::string>> matches;
  matches.clear();

  t1 = std::chrono::steady_clock::now();
  for (const auto& it : queries) {
    mih.queryAll(it.first, maxDistance, matches);
  }
  t2 = std::chrono::steady_clock::now();
  elapsedSeconds = t2 - t1;
  double seconds = elapsedSeconds.count();
```